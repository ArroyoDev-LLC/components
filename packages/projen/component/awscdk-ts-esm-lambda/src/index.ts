export * from './awscdk-ts-esm-lambda'
