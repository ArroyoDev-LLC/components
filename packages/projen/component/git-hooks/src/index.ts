export * from './git-hooks'
export * from './lint-staged'
