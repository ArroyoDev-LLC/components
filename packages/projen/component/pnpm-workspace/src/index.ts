export * from './pnpm-workspace.ts'
