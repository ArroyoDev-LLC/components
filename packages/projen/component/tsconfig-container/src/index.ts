export * from './tsconfig-container.ts'
