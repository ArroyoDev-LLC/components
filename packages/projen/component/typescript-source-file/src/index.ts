export * from './typescript-source-file'
export * from './typescript-source-config'
export * from './type-script-source-file-cache'
