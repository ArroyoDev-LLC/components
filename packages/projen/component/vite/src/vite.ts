import { LintConfig } from '@arroyodev-llc/projen.component.linting'
import {
	TypeScriptSourceFile,
	type TypeScriptSourceFileTransform,
} from '@arroyodev-llc/projen.component.typescript-source-file'
import {
	addPropertyAssignmentsFromObject,
	findComponent,
} from '@arroyodev-llc/utils.projen'
import { Component, DependencyType, type Project } from 'projen'
import { type NodePackage } from 'projen/lib/javascript'
import { type TypeScriptProject } from 'projen/lib/typescript'
import {
	type ImportDeclarationStructure,
	type ObjectLiteralExpression,
	type OptionalKind,
	type SourceFile,
	SyntaxKind,
	type WriterFunction,
} from 'ts-morph'
import type { UserConfigExport } from 'vite'

export interface ViteOptionsPlugin<OptionsT = Record<string, any>> {
	/**
	 * Name of plugin.
	 * This should match the import name from `moduleImport`.
	 */
	name: string
	/**
	 * Module import specification.
	 */
	moduleImport: OptionalKind<ImportDeclarationStructure>
	/**
	 * Plugin definition specification.
	 * Defaults to `<name>(<options>)`.
	 */
	spec?: WriterFunction | readonly (string | WriterFunction)[]
	/**
	 * Options object for plugin.
	 */
	options?: OptionsT
}

export interface ViteOptions {
	build: UserConfigExport
	plugins?: ViteOptionsPlugin[]
}

export class Vite extends Component {
	public static of(project: Project): Vite | undefined {
		return findComponent<typeof Vite>(project, Vite)
	}

	readonly file: TypeScriptSourceFile

	constructor(
		public readonly project: TypeScriptProject,
		public readonly options?: ViteOptions
	) {
		super(project)
		this.options = options ?? { build: {} }

		this.applyPackage(this.project.package)
		this.file = this.buildFile()
		this.addBuildConfig(this.options.build)
		this.project.tasks
			.tryFind('post-compile')
			?.exec?.('vite', { args: ['build'] })
		this.project
			.tryFindObjectFile('tsconfig.json')
			?.addToArray?.('compilerOptions.types', 'vite/client')
	}

	protected applyPackage(nodePackage: NodePackage): this {
		nodePackage.project.deps.addDependency('vite', DependencyType.BUILD)
		return this
	}

	protected buildFile(): TypeScriptSourceFile {
		const file = new TypeScriptSourceFile(this.project, 'vite.config.ts', {
			source: `export default defineConfig({})`,
			recreate: true,
		})
		file.addImport({
			moduleSpecifier: 'vite',
			namedImports: ['defineConfig'],
		})
		LintConfig.of(this.project)?.eslint?.addOverride?.({
			files: ['vite.config.ts'],
			rules: {
				'import/no-extraneous-dependencies': 'off',
			},
		})
		return file
	}

	/**
	 * Add Vite plugin to config.
	 * @param plugin Plugin specification.
	 */
	addPlugin(plugin: ViteOptionsPlugin): this {
		const { spec, name, moduleImport, options } = plugin
		const pluginSpec =
			spec ??
			((writer) =>
				writer.write(`${name}(${options ? JSON.stringify(options) : ''})`))

		this.project.deps.addDependency(
			moduleImport.moduleSpecifier.split('/')[0],
			DependencyType.BUILD
		)
		this.file.addImport(moduleImport)
		this.addConfigTransform((configExpr) => {
			const existsPlugins = configExpr
				.getProperty('plugins')
				?.asKind?.(SyntaxKind.PropertyAssignment)
			const pluginsExpr =
				existsPlugins?.asKindOrThrow(SyntaxKind.PropertyAssignment) ??
				configExpr.addPropertyAssignment({
					name: 'plugins',
					initializer: '[]',
				})
			pluginsExpr
				.getInitializerIfKindOrThrow(SyntaxKind.ArrayLiteralExpression)
				.addElements(pluginSpec)
		})
		return this
	}

	/**
	 * Add build config. Will be merged into existing.
	 * @param config Config to merge.
	 */
	addBuildConfig(config: UserConfigExport): this {
		this.addConfigTransform((configExpr) => {
			addPropertyAssignmentsFromObject(configExpr, config)
		})
		return this
	}

	/**
	 * Add `ts-morph` transformation on config node.
	 * @param transform Transformation to add.
	 */
	addConfigTransform(
		transform: (
			configObjectLiteral: ObjectLiteralExpression,
			sourceFile: SourceFile
		) => void
	): this {
		const transformer: TypeScriptSourceFileTransform = (src: SourceFile) => {
			const configExport = src
				.getExportAssignmentOrThrow((exp) =>
					Boolean(exp.getExpressionIfKind(SyntaxKind.CallExpression))
				)
				.getExpressionIfKindOrThrow(SyntaxKind.CallExpression)
			const configExpr = configExport
				.getArguments()[0]!
				.asKindOrThrow(SyntaxKind.ObjectLiteralExpression)
			transform(configExpr, src)
		}
		this.file.addTransformer(transformer)
		return this
	}
}
