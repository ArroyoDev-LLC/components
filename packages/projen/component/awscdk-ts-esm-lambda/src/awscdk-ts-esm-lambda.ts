import { existsSync } from 'node:fs'
import { TypeScriptSourceFile } from '@arroyodev-llc/projen.component.typescript-source-file'
import { findComponent, isComponent } from '@arroyodev-llc/utils.projen'
import {
	awscdk,
	Component,
	type FileBase,
	type Project,
	TextFile,
	typescript,
	javascript,
} from 'projen'

export interface AwsCdkTsEsmLambdaOptions {
	/**
	 * Project relative path to lambda function construct.
	 */
	readonly filePath: string
}

/**
 * AWS CDK TypeScript Lambda function with ESM support.
 */
export class AwsCdkTsEsmLambda extends Component {
	public static of(project: Project): AwsCdkTsEsmLambda | undefined {
		return findComponent(project, AwsCdkTsEsmLambda)
	}

	readonly lambdaFile: FileBase
	readonly tsFile: TypeScriptSourceFile

	constructor(project: Project, options: AwsCdkTsEsmLambdaOptions) {
		super(project)

		this.lambdaFile = project.tryFindFile(options.filePath) as FileBase

		const lambdaComponents = this.project.components.filter((comp) =>
			isComponent(awscdk.LambdaFunction, comp),
		)
		this.node.addDependency(...lambdaComponents.map((comp) => comp.node))

		const sourceCode = this.project.components.find(
			(comp) =>
				isComponent(TextFile, comp) && comp.path === this.lambdaFile.path,
		) as TextFile
		// @ts-expect-error would required replacing component or other post synth hackarounds otherwise since we need to replace the file before it gets written
		// eslint-disable-next-line @typescript-eslint/no-unsafe-assignment,@typescript-eslint/no-unsafe-call,@typescript-eslint/no-unsafe-member-access
		const source: string = sourceCode.lines.join('\n')
		this.project.tryRemoveFile(this.lambdaFile.path)

		if (!(this.project instanceof typescript.TypeScriptProject)) {
			throw new Error('AwsCdkTsEsmLambda must be used in a TypeScriptProject')
		}

		this.tsFile = new TypeScriptSourceFile(this.project, this.lambdaFile.path, {
			marker: false,
			recreate: true,
			source,
			format: false,
		})
		this.applyTransforms()
		javascript.Eslint.of(project)?.addIgnorePattern?.(this.tsFile.path)
	}

	/**
	 * Apply transforms to the lambda function source file.
	 * @protected
	 */
	protected applyTransforms() {
		this.tsFile.addImport({
			namedImports: ['fileURLToPath'],
			moduleSpecifier: 'node:url',
		})
		this.tsFile.addTransformer((sourceFile) => {
			const functionClass = sourceFile.getClasses()[0]
			if (!functionClass)
				throw new Error('Could not find class in entrypoint file')
			const functionCtor = functionClass.getConstructors()[0]
			if (!functionCtor)
				throw new Error('Could not find class constructor in entrypoint file')
			functionCtor.insertStatements(0, [
				'const __filename = fileURLToPath(import.meta.url);',
				'const __dirname = path.dirname(__filename);',
			])
			sourceFile.formatText()
		})
	}
}

/**
 * Automatically discover and create AWS CDK TypeScript Lambda functions with ESM support.
 */
export class AwsCdkTsEsmLambdaAutoDiscover extends Component {
	constructor(project: Project) {
		super(project)

		const lambdaFunctionFiles = project.files
			.filter((file) => file.path.endsWith('-function.ts'))
			.filter((file) =>
				existsSync(file.absolutePath.replace('-function.ts', '.lambda.ts')),
			)

		for (const lambdaFunctionFile of lambdaFunctionFiles) {
			new AwsCdkTsEsmLambda(project, { filePath: lambdaFunctionFile.path })
		}
	}
}
