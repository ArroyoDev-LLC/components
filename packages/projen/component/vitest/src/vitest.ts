import path from 'node:path'
import { TypeScriptSourceConfig } from '@arroyodev-llc/projen.component.typescript-source-file'
import { Vite } from '@arroyodev-llc/projen.component.vite'
import { cwdRelativePath } from '@arroyodev-llc/utils.projen'
import { Component, JsonFile, type Project, type Task } from 'projen'
import { type TypeScriptProject } from 'projen/lib/typescript'
import { deepMerge } from 'projen/lib/util'
import { type UserWorkspaceConfig } from 'vitest/config'

export enum VitestConfigType {
	WORKSPACE = 'Workspace',
	PROJECT = 'Project',
	CONFIG = 'Config',
}

export interface VitestOptions {
	/**
	 * Vitest config path.
	 */
	configFilePath?: string
	/**
	 * Configuration type.
	 */
	configType?: VitestConfigType
	/**
	 * Vitest settings.
	 */
	settings?: UserWorkspaceConfig
}

export class Vitest extends Component {
	public static of(project: Project): Vitest | undefined {
		const isVitest = (c: Component): c is Vitest => c instanceof Vitest
		return project.components.find(isVitest)
	}

	readonly configType: VitestConfigType
	readonly configFile: TypeScriptSourceConfig<UserWorkspaceConfig>
	readonly options: Required<VitestOptions>
	readonly vite?: Vite
	readonly testWatchTask: Task
	readonly #workspaceProjects: Map<string, Vitest> = new Map<string, Vitest>()

	constructor(
		public readonly project: TypeScriptProject,
		options: VitestOptions = {
			configFilePath: 'vitest.config.ts',
			configType: VitestConfigType.PROJECT,
			settings: {},
		}
	) {
		super(project)
		this.vite = Vite.of(this.project)
		this.options = deepMerge([
			{
				settings: {
					test: {
						name: project.name,
						include: [`${project.testdir}/\*\*/\*.spec.ts`],
					},
				},
			} as VitestOptions,
			options,
		]) as Required<VitestOptions>
		this.project.addDevDeps('vitest')
		this.configType = this.options.configType ?? VitestConfigType.PROJECT
		const configFilePath = this.options.configFilePath ?? 'vitest.config.ts'

		if (this.configType !== VitestConfigType.WORKSPACE) {
			this.project.testTask.exec('vitest', {
				args: ['--run'],
				receiveArgs: true,
			})
		}
		this.testWatchTask =
			this.project.tasks.tryFind('test:watch') ??
			this.project.addTask('test:watch', {
				receiveArgs: true,
				exec: 'vitest',
				description: 'Run tests on changes.',
			})

		this.project.tsconfigDev?.addInclude?.(configFilePath)
		this.project?.eslint?.addOverride?.({
			files: [
				configFilePath,
				...(this.options.settings?.test?.include ?? [this.project.testdir]),
			],
			rules: {
				'import/no-extraneous-dependencies': ['off'],
			},
		})

		this.configFile =
			TypeScriptSourceConfig.withCallExpressionConfig<UserWorkspaceConfig>(
				this.project,
				configFilePath,
				{
					source: `export default ${this.defineName}({})`,
					recreate: true,
					marker: true,
					config: this.options.settings,
				}
			)

		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: [this.defineName],
		})

		if (this.configType === VitestConfigType.PROJECT) {
			this.tryFindWorkspace()?.addProjectConfig?.(this)
		}

		if (this.configType === VitestConfigType.WORKSPACE) {
			new JsonFile(this.project, 'vitest.workspace.json', {
				readonly: true,
				allowComments: false,
				obj: () => Array.from(this.#workspaceProjects.keys()),
			})
		}
	}

	/**
	 * Get the define function name.
	 * @protected
	 */
	protected get defineName(): string {
		return this.configType === VitestConfigType.PROJECT
			? 'defineProject'
			: 'defineConfig'
	}

	/**
	 * Merge vite config if applicable.
	 * @protected
	 */
	protected mergeVite(): void {
		if (!this.vite) return
		this.configFile.addImport({
			moduleSpecifier: 'vitest/config',
			namedImports: ['mergeConfig'],
		})
		this.configFile.addImport({
			moduleSpecifier: cwdRelativePath(
				this.project.outdir,
				this.vite.file.absolutePath.split('.').slice(0, -1).join('.')
			),
			defaultImport: 'viteConfig',
		})
		this.configFile.addConfigTransform((cfg, src) => {
			src.addExportAssignment({
				expression: (writer) =>
					writer.write(
						`mergeConfig(viteConfig, ${this.defineName}(${cfg.getFullText()}))`
					),
				isExportEquals: false,
			})
			src.removeDefaultExport()
		})
	}

	/**
	 * Try to find the workspace {@link Vitest} component.
	 */
	tryFindWorkspace(): Vitest | undefined {
		if (this.configType === VitestConfigType.WORKSPACE) return this
		if (this.project.parent) {
			return Vitest.of(this.project.parent)?.tryFindWorkspace?.()
		}
		return
	}

	/**
	 * Add {@link Vitest} component to this workspace.
	 * @param component Target component.
	 */
	addProjectConfig(component: Vitest): this {
		if (this.configType !== VitestConfigType.WORKSPACE)
			throw new Error('No vitest workspace found. Cannot add project config!')
		const relPath = path.format({
			dir: '.',
			base: path.relative(
				path.dirname(this.configFile.absolutePath),
				component.configFile.absolutePath
			),
		})
		this.#workspaceProjects.set(relPath, component)
		return this
	}

	/**
	 * Add config to merge.
	 * @param config Config part.
	 */
	addConfig(config: this['configFile']['__mergeSchema']): this {
		this.configFile.addConfig(config)
		return this
	}

	/**
	 * Add plugin to merge.
	 * @param plugin Plugin part.
	 */
	addPlugin(plugin: this['configFile']['__pluginSchema']) {
		this.configFile.addPlugin(plugin)
		return this
	}

	/**
	 * @inheritDoc
	 */
	preSynthesize() {
		super.preSynthesize()
		this.mergeVite()
	}
}
