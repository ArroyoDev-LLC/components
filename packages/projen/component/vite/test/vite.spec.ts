import os from 'node:os'
import path from 'node:path'
import { typescript } from 'projen'
import { Testing } from 'projen/lib/testing'
import { beforeEach, expect, test } from 'vitest'
import { Vite } from '../src'

interface TestContext {
	project: typescript.TypeScriptProject
}

beforeEach<TestContext>((ctx) => {
	ctx.project = new typescript.TypeScriptProject({
		name: ctx.meta.id,
		outdir: path.join(os.tmpdir(), ctx.meta.id),
		defaultReleaseBranch: 'main',
	})
})

test<TestContext>('renders as expected', async (ctx) => {
	new Vite(ctx.project)
	const synth = Testing.synth(ctx.project)
	expect(synth['vite.config.ts']).toBeDefined()
	expect(synth['vite.config.ts']).toMatchSnapshot()
	expect(synth['package.json']).toMatchSnapshot()
})

test<TestContext>('renders as expected with merged config', async (ctx) => {
	const vite = new Vite(ctx.project, {
		build: {
			resolve: {
				dedupe: ['vue'],
			},
		},
	})
	vite.addBuildConfig({
		resolve: {
			alias: {
				'@': (writer) => writer.write('fileUrlToPath(new URL())'),
			},
			dedupe: ['other'],
		},
	})
	const synth = Testing.synth(ctx.project)
	expect(synth['vite.config.ts']).toBeDefined()
	expect(synth['vite.config.ts']).toMatchSnapshot()
})

test<TestContext>('renders with plugins as expected', async (ctx) => {
	const vite = new Vite(ctx.project)
	vite.addPlugin({
		name: 'vueJsx',
		moduleImport: {
			moduleSpecifier: '@vite/plugin-vue-jsx',
			defaultImport: 'vueJsx',
		},
	})
	const synth = Testing.synth(ctx.project)
	expect(synth['vite.config.ts']).toMatchSnapshot()
})
