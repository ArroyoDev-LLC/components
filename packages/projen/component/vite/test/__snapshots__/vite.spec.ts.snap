// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`renders as expected 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

import { defineConfig } from \\"vite\\";

export default defineConfig({})"
`;

exports[`renders as expected 2`] = `
{
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".",
  "devDependencies": {
    "@types/jest": "*",
    "@types/node": "^16",
    "@typescript-eslint/eslint-plugin": "^6",
    "@typescript-eslint/parser": "^6",
    "eslint": "^8",
    "eslint-import-resolver-node": "*",
    "eslint-import-resolver-typescript": "*",
    "eslint-plugin-import": "*",
    "jest": "*",
    "jest-junit": "^15",
    "npm-check-updates": "^16",
    "projen": "*",
    "standard-version": "^9",
    "ts-jest": "*",
    "typescript": "*",
    "vite": "*",
  },
  "jest": {
    "clearMocks": true,
    "collectCoverage": true,
    "coverageDirectory": "coverage",
    "coveragePathIgnorePatterns": [
      "/node_modules/",
    ],
    "coverageReporters": [
      "json",
      "lcov",
      "clover",
      "cobertura",
      "text",
    ],
    "globals": {
      "ts-jest": {
        "tsconfig": "tsconfig.dev.json",
      },
    },
    "preset": "ts-jest",
    "reporters": [
      "default",
      [
        "jest-junit",
        {
          "outputDirectory": "test-reports",
        },
      ],
    ],
    "testMatch": [
      "<rootDir>/src/**/__tests__/**/*.ts?(x)",
      "<rootDir>/(test|src)/**/*(*.)@(spec|test).ts?(x)",
    ],
    "testPathIgnorePatterns": [
      "/node_modules/",
    ],
    "watchPathIgnorePatterns": [
      "/node_modules/",
    ],
  },
  "license": "Apache-2.0",
  "main": "lib/index.js",
  "name": "-884671080_0",
  "scripts": {
    "build": "npx projen build",
    "bump": "npx projen bump",
    "clobber": "npx projen clobber",
    "compile": "npx projen compile",
    "default": "npx projen default",
    "eject": "npx projen eject",
    "eslint": "npx projen eslint",
    "package": "npx projen package",
    "post-compile": "npx projen post-compile",
    "post-upgrade": "npx projen post-upgrade",
    "pre-compile": "npx projen pre-compile",
    "projen": "npx projen",
    "release": "npx projen release",
    "test": "npx projen test",
    "test:watch": "npx projen test:watch",
    "unbump": "npx projen unbump",
    "upgrade": "npx projen upgrade",
    "watch": "npx projen watch",
  },
  "types": "lib/index.d.ts",
  "version": "0.0.0",
}
`;

exports[`renders as expected with merged config 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

import { defineConfig } from \\"vite\\";

export default defineConfig({
    resolve: {
        dedupe: [\\"vue\\",  \\"other\\"],
        alias: {
            \\"@\\": fileUrlToPath(new URL())
        }
    }
})"
`;

exports[`renders with plugins as expected 1`] = `
"// ~~ Generated by projen. To modify, edit .projenrc.js and run \\"npx projen\\".

import vueJsx from \\"@vite/plugin-vue-jsx\\";
import { defineConfig } from \\"vite\\";

export default defineConfig({
    plugins: [vueJsx()]
})"
`;
